tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,504 @@ keywords:
+
+Then we can see that merge sort is performed in this way. 
+
+![image](https://miro.medium.com/max/600/1*opwN0BhtH4zvPF697fPlow.gif) 
+
+Image by Brian Hans via [Medium](https://medium.com/@brianhans/merge-sort-a1d031eaa40f)
+
+## Algorithm 
+### Divide (Dividing is our first step)
+- Calculate the midpoint by checking if the left index is less than the right index, if yes divide the array. 
+- Now continue dividing the array until `left index < right index` becomes false, that is until the division is not possible.
+     
+### Conquer
+- After dividing the array into the smallest units, start merging the elements again by comparing them.
+- We need to compare and merge starting from the last splits or last smallest units. So **Recursion** needs to be done here.
+### Merge 
+- Since each half is already sorted so we need to just sort between 2 halves to **combine/merge** than to make a bigger sorted array.  
+## [Leetcode Problem 912 Sort an Array](https://leetcode.com/problems/sort-an-array/)
+
+**Problem Statement** 
+
+Given an array of integers nums, sort the array in ascending order and return it.
+You must solve the problem without using any built-in functions in O(nlog(n)) time complexity and with the smallest space complexity possible.
+
+**Test Cases**
+    
+Test case 1 :
+     
+```
+Input: nums = [5,2,3,1]
+Output: [1,2,3,5]
+Explanation: 
+    After sorting the array, the positions of some numbers are not changed (for example, 2 and 3). 
+    While the positions of other numbers are changed (for example, 1 and 5).       
+```
+**Constraints as given in the problem:**
+```
+1 <= nums.length <= 5 * 10^4
+-5 * 104 <= nums[i] <= 5 * 10^4
+```
+*Here `insertion sort`, `quick sort`, `selection sort` and all other sorts in $O(n^2)$ will give **TLE** i.e Time Limit Exceeded*
+
+## STL Approach  
+Note: Not preferred in this problem since it is a built-in function, but still you will get an AC in Leetcode.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+void stableStlSort(vector<int>& nums){
+    stable_sort(nums.begin(), nums.end());
+}
+```
+</TabItem>
+
+<TabItem value="py" label="Python">
+
+```py
+def sortArray(self, nums):
+    return sorted(nums)
+```
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+public int[] sortArray(int[] nums) {
+    Arrays.sort(nums);
+    return nums;
+}
+```
+</TabItem>
+</Tabs>
+
+## Top Down Approach
+### Algorithm    
+```
+MergeSort(array[], left,  right)
+If right > left
+1. Find the middle point to divide the array into two halves:  
+    middle = left + (right – left)/2
+2. Call mergeSort again for first half for further dividing:   
+    Call mergeSort(array, left, middle)
+3. Call mergeSort again  for second half for further dividing:
+    Call mergeSort(array, middle+ 1, right)
+4. Merge the two halves sorted :
+    Call merge(arrary, left, middle, right)
+```
+<Tabs>
```

Comment:
1. i expect to see same detailed comments in 3 different langs. 
2. adopt in LC solution format. Don't just show 2 functions. (i.e. i can copy and paste to run in LC directly)

Reply:
@wingkwong  this is the generalized algorithm for merge sort, So how could I add comments in different languages and introduce LC format?